cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -640,8 +641,21 @@ private void executeAll(Iterator<WorkflowNode> iterator, ApplicationSpecificatio
           workflowContext.setState(new ProgramState(ProgramStatus.KILLED, rootCause.getMessage()));
           break;
         }
-        workflowContext.setState(new ProgramState(ProgramStatus.FAILED, rootCause.getMessage()));
-        throw Throwables.propagate(rootCause);
+        StringBuilder contextFailMessage = new StringBuilder();
```

Comment:
Can you refactor this logic into some helper function? Maybe you can add a `Exceptions` util class in cdap-common under the `io.cdap.cdap.common.lang` package.

Reply:
Done